kernel/kernel-api/src/main/java/io/delta/kernel/types/FieldMetadata.java

```diff
@@ -0,0 +1,203 @@
+    }
+
+    public String toJson() {
+        return metadata.entrySet().stream()
+            .map(e -> String.format("\"%s\" : %s", e.getKey(), valueToJson(e.getValue())))
+            .collect(Collectors.joining(",\n", "{", "}"));
+    }
+
+    // TODO this is a hack for now until we have real serialization as part of the JsonHandler
+    /** Wraps string objects in quotations, otherwise converts using toString() */
+    private String valueToJson(Object value) {
+        if (value == null) {
+            return "null";
+        } else if (value instanceof String) {
+            return String.format("\"%s\"", value);
+        } else {
+            return value.toString();
+        }
+    }
+
+    @Override
+    public String toString() {
+        return toJson();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        FieldMetadata that = (FieldMetadata) o;
+        if (this.metadata.size() != that.metadata.size()) return false;
+        return this.metadata.entrySet().stream().allMatch(e ->
+            Objects.equals(e.getValue(), that.metadata.get(e.getKey())) ||
+                (e.getValue() != null && e.getValue().getClass().isArray() &&
+                    that.metadata.get(e.getKey()).getClass().isArray() &&
+                    Arrays.equals(
+                        (Object[]) e.getValue(),
+                        (Object[]) that.metadata.get(e.getKey()))));
+    }
+
+    @Override
+    public int hashCode() {
+        return metadata.entrySet()
+            .stream()
+            .mapToInt( entry -> (entry.getValue().getClass().isArray() ?
+                (entry.getKey() == null ? 0 : entry.getKey().hashCode())^
+                    Arrays.hashCode((Object[]) entry.getValue()) :
+                entry.hashCode())
+            ).sum();
+    }
+
+    /**
+     * @return a new {@link FieldMetadata.Builder}
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Builder class for {@link FieldMetadata}.
+     */
+    public static class Builder {
+        private Map<String, Object> metadata = new HashMap<String, Object>();
+
+        public Builder putNull(String key) {
+            metadata.put(key, null);
+            return this;
+        }
+
+        public Builder putLong(String key, long value) {
+            metadata.put(key, value);
+            return this;
+        }
+
+        public Builder putDouble(String key, double value) {
+            metadata.put(key, value);
+            return this;
+        }
+
+        public Builder putBoolean(String key, boolean value) {
+            metadata.put(key, value);
+            return this;
+        }
+
+        public Builder putString(String key, String value) {
+            metadata.put(key, value);
+            return this;
+        }
+
+        public Builder putMetadata(String key, FieldMetadata value) {
```

Comment:
what's a `Metadata` ? Would `putFieldMetadata` be clearer?

Reply:
Updated